### ISSUE TITLE
NPE in IncompatibleLinkageProblem

### ISSUE THREAD

**Issue by marquiswang** [GoogleCloudPlatform/cloud-opensource-java#1599]
I'm getting the following exception when I run the linkage-checker version 1.5.0.

I can try too make a minimum reproducing pom if necessary.

```
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException
    at com.google.cloud.tools.opensource.classpath.IncompatibleLinkageProblem.formatSymbolProblem (IncompatibleLinkageProblem.java:36)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:708)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.LongPipeline.reduce (LongPipeline.java:438)
    at java.util.stream.LongPipeline.sum (LongPipeline.java:396)
    at java.util.stream.ReferencePipeline.count (ReferencePipeline.java:526)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:221)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:205)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by marquiswang** [comment]
This appears to already be fixed in 1.5.1-SNAPSHOT, so perhaps no action is needed (though a release would be appreciated).

**Comment by suztomo** [comment]
@marquiswang Thank you for reporting and checking with the snapshot. Can I get the pom.xml? I'll add test to ensure this won't happen again.

Memo: the stacktrace indicates the first line of formatSymbolProblem. `targetClass` is not supposed to be null.

```
  @Override
  public String formatSymbolProblem() {
    String jarInfo = "(" + targetClass.getClassPathEntry() + ") ";
    return jarInfo + super.formatSymbolProblem();
  }
```

The IncompatibleLinkageProblem class has been removed today from master but not because of the NPE.

**Comment by elharo** [comment]
Completly coincidentally I did add a null check there though so this is probably already fixed at head.

**Comment by marquiswang** [comment]
@suztomo it looks like it is caused by the following dependency:

```
<dependency>
<groupId>com.sun</groupId>
<artifactId>tools</artifactId>
<version>1.7</version>
<scope>system</scope>
<systemPath>${java.home}/../lib/tools.jar</systemPath>
</dependency>
```

**Comment by suztomo** [comment]
@marquiswang Thank you. I'm able to reproduce the problem. (Initially I thought the version 1.7 was the culprit, but even with `1.8`, Linkage Checker fails)

https://gist.github.com/suztomo/f8e51d4e1090733aa75e21f1010f1d47

**Comment by suztomo** [comment]
Memo:

The problematic reference is from `com.sun.tools.internal.ws.wscompile.WsgenOptions` (source) referencing `com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl` (target). Linkage Checker can load the latter through `classDumper.loadJavaClass` but it cannot get the JAR (`tools.jar`) from `classDumper.findClassLocation`.

The class path here was 
`/Users/suztomo/i1599/target/linkage-checker-test-1.0-SNAPSHOT.jar` and `/Library/Java/JavaVirtualMachines/jdk1.8.0_221.jdk/Contents/Home/jre/../lib/tools.jar`.

**Comment by elharo** [comment]
https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1594

**Comment by elharo** [comment]
It sounds like the linkage error is a true positive? If so, should we close this?

**Comment by marquiswang** [comment]
Thank you for your help @suztomo!

@elharo I think that the NPE was still a bug, though if you've already added the null check then I think we're good.

**Comment by suztomo** [comment]
We released Linkage Checker enforcer rule 1.5.1 that resolves this NPE bug.

As you see, there are other issues/pull requests we're working on. I'll keep you posted on the progress.
